File: android/app/src/main/kotlin/com/abstractvision/phonedetox/foreground/InterventionPackagePolicy.kt
package com.abstractvision.phonedetox.foreground

object InterventionPackagePolicy {
    private const val PHONE_DETOX_PACKAGE = "com.abstractvision.phonedetox"
    private val exactSafePackages = setOf("com.android.systemui", "com.android.settings")
    private val safeParts = setOf(
        "permissioncontroller",
        "packageinstaller",
        "incallui",
        "telecom",
        "emergency",
        "setupwizard",
        "systemui",
        "recovery",
    )

    fun isConfigurable(packageName: String): Boolean =
        packageName.isNotBlank() &&
            packageName != PHONE_DETOX_PACKAGE &&
            packageName !in exactSafePackages &&
            safeParts.none { packageName.contains(it, ignoreCase = true) }

    fun isTransientOrCritical(
        packageName: String,
        ownPackageName: String,
        dynamicSafePackages: Set<String>,
    ): Boolean =
        packageName == ownPackageName ||
            packageName in dynamicSafePackages ||
            packageName in exactSafePackages ||
            safeParts.any { packageName.contains(it, ignoreCase = true) }
}

File: lib/core/domain/intervention_package_policy.dart
abstract final class InterventionPackagePolicy {
  static bool isConfigurable(String packageName) {
    final value = packageName.trim().toLowerCase();
    if (value.isEmpty || value == 'com.abstractvision.phonedetox') return false;
    return !_excludedParts.any(value.contains);
  }

  static const _excludedParts = <String>{
    'systemui',
    'settings',
    'permissioncontroller',
    'packageinstaller',
    'incallui',
    'telecom',
    'emergency',
    'setupwizard',
    'recovery',
  };
}
